### PR TITLE
iOS TextSelectionToolbar fidelity

### PR DESCRIPTION
**Pull request by luccasclezar**
CupertinoTextSelectionToolbar is different from the native one, with some UI and UX issues. More details on the linked issue.

https://github.com/flutter/flutter/issues/127756

Currently the only problem that I listed on the linked issue that I couldn't fix was the horizontal scrolling, but to workaround this I added a GestureDetector to change pages when swiping the toolbar. It's not exactly the same as native as there is no scroll animation, but it works.

I'm creating this PR a little early to have some feedback as these changes were more complex than the ones in my last PR. Probably best if @justinmc is involved 😅

|Version|Video|
|-|-|
|Flutter Old|<video src="https://github.com/flutter/flutter/assets/12024080/7cf81075-46ec-4970-b118-cc27b60ddac0"></video>|
|Flutter New|<video src="https://github.com/flutter/flutter/assets/12024080/c9e27a53-f94c-4cb0-9b76-e47b73841dcb"></video>|
|Native|<video src="https://github.com/flutter/flutter/assets/12024080/468c7d5b-ba93-4bd4-8f6e-8ec2644b9866"></video>|

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/cupertino/text_selection_toolbar.dart [hunk 1]
```diff
@@ -903,6 +940,11 @@ class _RenderCupertinoTextSelectionToolbarItems extends RenderBox with Container
         // already been taken care of when laying out the children to
         // accommodate the back button.
       }
+
+      // Update the controller values so that we can check in the horizontal
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:944]: I think the [Flutter style guide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo) recommends not using "we" etc., so maybe say "...so that it's possible to check them in the horizontal..."

**Review by justinmc — COMMENTED** [review]
This is an awesome PR, the menu looks so much better from your videos.  I think there's one architectural problem to solve (_CupertinoTextSelectionToolbarItemsController) but otherwise the approach looks good.

Is the point of that controller to be able to pass information up the class hierarchy from _RenderCupertinoTextSelectionToolbarItems  to _CupertinoTextSelectionToolbarContentState?  I think we need to find a better way to do that other than a shared mutable class.

Fortunately it seems like you only need that information in a callback (_handleNext/PreviousPage) so we can probably look it up without having to worry about layout order.  Could you pass a key to _CupertinoTextSelectionToolbarItems and then look up the information using GlobalKey.currentWidget?  Something like that should be doable...

**Comment by luccasclezar** [comment]
Thank you! 😁

About the controller, yes, I only use it to know if it's possible to navigate to previous/next page when dragging, basically.

But I don't understand how a GlobalKey would solve this problem. The only way to know if there is a previous/next button (as far as I can tell) is in `_RenderCupertinoTextSelectionToolbarItems.performLayout`:

https://github.com/flutter/flutter/blob/6762521feb2821287b66bbd315cfe3b2fba56857/packages/flutter/lib/src/cupertino/text_selection_toolbar.dart#L946-L947

So we need to set the variables in this method. From what I understood of your suggestion, a GlobalKey should be used instead of the "controller", like 

```dart
globalKey.currentWidget.hasNextPage = page != currentPage;
globalKey.currentWidget.hasPreviousPage = page > 0; 
```

But how would I set variables using `globalKey.currentWidget` if all widget variables are final? If `_RenderCupertinoTextSelectionToolbarItems` were a Widget I could assign it a GlobalKey and get its properties, but RenderBox doesn't have a `key` parameter.

Probably I'm missing something here, but I couldn't come up with a solution that doesn't involve a shared mutable class. 😅

File: packages/flutter/lib/src/cupertino/text_selection_toolbar.dart [hunk 2]
```diff
@@ -510,49 +520,77 @@ class _CupertinoTextSelectionToolbarContentState extends State<_CupertinoTextSel
...
     ));
   }
 }
 
+// This doesn't really control the toolbar, but has updated values to check if
+// there are previous/next pages. This wasn't needed before because the toolbar
+// correctly shows/hides the previous/next buttons automatically, but now the
+// horizontal drag gesture callback has to check whether it's possible to
+// navigate or not, making this necessary.
+class _CupertinoTextSelectionToolbarItemsController {
```
Comment (tgucio) [inline packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:562]: I think controller classes are normally used in APIs for others widgets to change the state of a stateful widget. Here we're operating on a render object so this should probably be done using individual fields e.g.

```
void updateRenderObject(BuildContext context, _RenderCupertinoTextSelectionToolbarItems renderObject) {
    renderObject
      ..page = page
      ..hasPreviousPage = hasPreviousPage
      ..hasNextPage = hasNextPage
      ..dividerColor = dividerColor
```

Then in the `RenderBox` class you'd have a setter and a getter just like with `dividerColor`.

Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:562]: The problem with this approach is that, as far as I understand it, I wouldn't be able to update `hasNextPage` and `hasPreviousPage` in `_RenderCupertinoTextSelectionToolbarItems`, and have it available in its parent (`_CupertinoTextSelectionToolbarContentState`) when the user does a horizontal drag gesture on the whole toolbar.

This is where I update the values:
https://github.com/flutter/flutter/blob/6762521feb2821287b66bbd315cfe3b2fba56857/packages/flutter/lib/src/cupertino/text_selection_toolbar.dart#L944-L947

And this is where I use them:
https://github.com/flutter/flutter/blob/6762521feb2821287b66bbd315cfe3b2fba56857/packages/flutter/lib/src/cupertino/text_selection_toolbar.dart#L461-L475

Comment (tgucio) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:562]: If the `hasPreviousPage` / `hasNextPage` fields are only updated by `_RenderCupertinoTextSelectionToolbarItems` (which makes sense given this will depend on the incoming constraints for the RenderBox), perhaps one approach would be to:
- keep them as `bool` values without getter/setter
- from the widget state, in the gesture handler, do something like:
```
final RenderBox? renderBox = context.findRenderObject() as RenderBox?;
if (renderBox is _RenderCupertinoTextSelectionToolbarItems) {
  if (renderBox.hasPreviousPage...)

```

Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:562]: That makes sense and it's way cleaner, thank you! I will see if I can remove the controller until this weekend.

Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart]: Just fixed this and all tests passed, should be OK now 😁

File: packages/flutter/lib/src/cupertino/text_selection_toolbar_button.dart [hunk 3]
```diff
@@ -89,6 +85,10 @@ class CupertinoTextSelectionToolbarButton extends StatelessWidget {
   /// {@endtemplate}
   final String? text;
 
+  /// Whether the button is enabled or disabled. Buttons are disabled by default. To
+  /// enable a button, set its [onPressed] property to a non-null value.
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/text_selection_toolbar_button.dart:89]: The first sentence should be it's own paragraph according to the styleguide.  So put an empty `///` line between the first sentence and the rest.

Comment (justinmc) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar_button.dart:89]: Actually, it looks like this property isn't used outside of the `State` below.  Can you make it private?

File: packages/flutter/lib/src/cupertino/text_selection_toolbar.dart [hunk 4]
```diff
@@ -449,19 +442,43 @@ class _CupertinoTextSelectionToolbarContent extends StatefulWidget {
...
+  }
 
   void _handleNextPage() {
-    _controller.reverse();
-    _controller.addStatusListener(_statusListener);
-    _nextPage = _page + 1;
+    final RenderBox? renderToolbar =
+      _toolbarItemsKey.currentContext?.findRenderObject() as RenderBox?;
+
+    if (renderToolbar is _RenderCupertinoTextSelectionToolbarItems && renderToolbar.hasNextPage) {
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:466]: This looks good now, thanks!

File: packages/flutter/lib/src/cupertino/text_selection_toolbar.dart (hunk 1 above)
Comment (justinmc) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:938]: Reminder on this one.

**Review by justinmc — COMMENTED** [review]
The GlobalKey approach looks good.  Just one comment about a public property that I think can be private, otherwise this looks good to merge.  Thank you for the fixes!

File: packages/flutter/lib/src/cupertino/text_selection_toolbar_button.dart (hunk 3 above)
Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar_button.dart:89]: Actually, there's no need to make it private, I added this property when I tried removing `CupertinoButton` from `CupertinoTextSelectionToolbarButton` widget, but I ended up re-adding `CupertinoButton`, so this property is not needed anymore.

File: packages/flutter/lib/src/cupertino/text_selection_toolbar.dart [hunk 5]
```diff
@@ -28,15 +29,29 @@ const double _kArrowScreenPadding = 26.0;
 // Values extracted from https://developer.apple.com/design/resources/.
 const Radius _kToolbarBorderRadius = Radius.circular(8);
 
+// Color was measured from a screenshot of iOS 16.0.2
+// TODO(LongCatIsLooong): https://github.com/flutter/flutter/issues/41507.
+const CupertinoDynamicColor _kToolbarBackgroundColor =
+    CupertinoDynamicColor.withBrightness(
+  color: Color(0xFFF6F6F6),
+  darkColor: Color(0xFF222222),
+);
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:38]: Nit: I would indent these 3 lines by 4 spaces.

**Review by justinmc — APPROVED** [review]
LGTM 👍 

@tgucio Can you give the secondary review here?

File: packages/flutter/lib/src/cupertino/text_selection_toolbar.dart [hunk 6]
```diff
@@ -117,29 +125,14 @@ class CupertinoTextSelectionToolbar extends StatelessWidget {
...
   // background and a rounded cutout with an arrow.
   static Widget _defaultToolbarBuilder(BuildContext context, Offset anchor, bool isAbove, Widget child) {
     final Widget outputChild = _CupertinoTextSelectionToolbarShape(
       anchor: anchor,
       isAbove: isAbove,
-      child: DecoratedBox(
-        decoration: BoxDecoration(
-          color: _kToolbarDividerColor.resolveFrom(context),
-        ),
+      child: ColoredBox(
```
Comment (tgucio) [inline packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:134]: Good catch turning this into a `ColoredBox`!

File: packages/flutter/lib/src/cupertino/text_selection_toolbar.dart [hunk 7]
```diff
@@ -484,7 +501,7 @@ class _CupertinoTextSelectionToolbarContentState extends State<_CupertinoTextSel
       value: 1.0,
       vsync: this,
       // This was eyeballed on a physical iOS device running iOS 13.
-      duration: const Duration(milliseconds: 150),
+      duration: _kToolbarTransitionDuration,
```
Comment (tgucio) [inline packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:504]: Also happy to see a constant here :)

File: packages/flutter/lib/src/cupertino/text_selection_toolbar_button.dart [hunk 8]
```diff
@@ -114,25 +110,60 @@ class CupertinoTextSelectionToolbarButton extends StatelessWidget {
...
+         style: _kToolbarButtonFontStyle.copyWith(
+           color: widget.onPressed != null
+               ? _kToolbarTextColor.resolveFrom(context)
+               : CupertinoColors.inactiveGray,
+         ),
+       ),
     );
+
+    if (widget.onPressed != null) {
+      // The toolbar buttons on iOS are different from the default CupertinoButton.
```
Comment (tgucio) [inline packages/flutter/lib/src/cupertino/text_selection_toolbar_button.dart:157]: Nit: this comment is on colours but it doesn't seem to be related to the code (just a `GestureDetector`?

**Review by tgucio — CHANGES_REQUESTED** [review]

File: packages/flutter/lib/src/cupertino/text_selection_toolbar.dart [hunk 9]
```diff
@@ -510,21 +527,35 @@ class _CupertinoTextSelectionToolbarContentState extends State<_CupertinoTextSel
...
+        curve: Curves.decelerate,
+        child: GestureDetector(
+          onHorizontalDragEnd: _onHorizontalDragEnd,
+          child: _CupertinoTextSelectionToolbarItems(
+            key: _toolbarItemsKey,
+            page: _page,
+            backButton: CupertinoTextSelectionToolbarButton(
+              onPressed: _handlePreviousPage,
+              child: Icon(
+                CupertinoIcons.chevron_left,
```
[the remark this answers is not in the document]
Comment (tgucio) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:541]: Ah, looks like for 2. we may also be able to override mirroring in RTL by simply supplying `textDirection` to `Icon`>

[the remark this answers is not in the document]
Comment (justinmc) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:541]: For 1, are you worried about increasing app size?  I see that CupertinoIcons is used elsewhere in Cupertino, like [in CupertinoTextField](https://github.com/flutter/flutter/blob/ca495da92a18a4b8bcd7f7953a08cbff73153832/packages/flutter/lib/src/cupertino/text_field.dart#L1195).  @MitchellGoodwin do you know if there's any concern with using CupertinoIcons like this?

For 2, good call.  @luccasclezar should try it out and make sure it looks right in RTL.

[the remark this answers is not in the document]
Comment (MitchellGoodwin) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:541]: Yes, CupertinoIcons are used throughout, including CupertinoApp itself, so I do not think it will increase the size at all. Only concerns with the package is that it's somewhat out of date, but the chevron icons aren't an issue.

[the remark this answers is not in the document]
Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:541]: About 2, I actually noticed this, but I'm using `CupertinoIcons.chevron_left` (and right), **not** `CupertinoIcons.left_chevron`. The only difference between the two if you look at the source code is that `chevron_left` does not have `matchTextDirection` enabled.

https://github.com/luccasclezar/flutter/blob/a36b1f9912de66a2588a7e9bababc6133b29fc31/packages/flutter/lib/src/cupertino/icons.dart#L1767

Regardless of the theory, I ran Flutter Gallery, enabled "Force RTL" and added a bunch of "Select All" items to the toolbar, this is the result:

Screenshot_20230606_211838

File: packages/flutter/lib/src/cupertino/text_selection_toolbar.dart (hunk 5 above)
Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:38]: I could format this as the other `CupertinoDynamicColor`s are formatted too, like this:

```dart
const CupertinoDynamicColor _kToolbarBackgroundColor = CupertinoDynamicColor.withBrightness(
  color: Color(0xFFF6F6F6),
  darkColor: Color(0xFF222222),
);
```

What do you think?

Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:38]: I ended up doing what @tgucio suggested as it was already how the other `CupertinoDynamicColor`s were formatted.

File: packages/flutter/lib/src/cupertino/text_selection_toolbar.dart [hunk 10]
```diff
@@ -920,6 +956,18 @@ class _RenderCupertinoTextSelectionToolbarItems extends RenderBox with Container
       if (childParentData.shouldPaint) {
         final Offset childOffset = childParentData.offset + offset;
         context.paintChild(child, childOffset);
+
+        // backButton is not in the children list, that's probably why its
```
[the remark this answers is not in the document]
Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:960]: I'm 99% sure that of that, so I can ~probably~ remove the "probably" 😅

Maybe @justinmc can verify that this comment is right?

File: packages/flutter/lib/src/cupertino/text_selection_toolbar_button.dart [hunk 11]
```diff
@@ -114,25 +110,60 @@ class CupertinoTextSelectionToolbarButton extends StatelessWidget {
...
       borderRadius: null,
-      color: _kToolbarBackgroundColor,
-      disabledColor: _kToolbarBackgroundColor,
-      onPressed: onPressed,
+      disabledColor: const Color(0x00000000),
+      onPressed: widget.onPressed,
       padding: _kToolbarButtonPadding,
-      pressedOpacity: onPressed == null ? 1.0 : 0.7,
-      child: child,
+      // There's no foreground fade on iOS toolbar anymore, it just darkens the
```
[the remark this answers is not in the document]
Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar_button.dart:142]: `CupertinoColors.inactiveGray` is a `CupertinoDynamicColor` so I thought it would stay OK in dark mode too, but I tested anyway :)

Screenshot_20230606_215359

File: packages/flutter/lib/src/cupertino/text_selection_toolbar_button.dart (hunk 8 above)
Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar_button.dart:157]: I improved the comments in this build method.

File: packages/flutter/lib/src/cupertino/text_selection_toolbar.dart (hunk 9 above)
[the remark this answers is not in the document]
Comment (tgucio) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:540]: @justinmc @MitchellGoodwin wouldn't this require adding the `cupertino_icons` dependency? Note that the Cupertino toolbar is also used for the regular `TextField` e.g. in a `MaterialApp`.

[the remark this answers is not in the document]
Comment (MitchellGoodwin) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:540]: The Cupertino icons is a static internal package that was generated based on an older version of the outside package `cupertino_icons`. So Flutter comes with them already ready to go, and we need to manually update the Cupertino icons to catch changes to `cupertino_icons` so we aren't in danger of breaking changes from that package.

File: packages/flutter/lib/src/cupertino/text_selection_toolbar.dart [hunk 12]
```diff
@@ -920,6 +955,18 @@ class _RenderCupertinoTextSelectionToolbarItems extends RenderBox with Container
       if (childParentData.shouldPaint) {
         final Offset childOffset = childParentData.offset + offset;
         context.paintChild(child, childOffset);
+
+        // backButton is not in the children list, that's why its
+        // childParentData.nextSibling is null. So either when there's a
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:960]: "backButton is a slotted child and is not in the children list, so its childParentData.nextSibling is null. ..."

File: packages/flutter/lib/src/cupertino/text_selection_toolbar.dart (hunk 10 above)
[the remark this answers is not in the document]
Comment (justinmc) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart]: That's right!  backButton, nextButton, and nextButtonDisabled are "slotted" children.  They have their own names and don't appear in the `children` list.  I'll leave another comment with a nit on the wording here but overall it is correct now.

File: packages/flutter/lib/src/cupertino/text_selection_toolbar.dart (hunk 9 above)
[the remark this answers is not in the document]
Comment (justinmc) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:540]: @MitchellGoodwin What about for apps that don't use Cupertino at all, will the icons be included in the bundle?  If so, then I believe that someone that isn't using Cupertino at all, but does use a TextField, will now have the Cupertino icons included due to this PR's change.

I wouldn't be surprised if this PR doesn't actually impact this in practice, but it would be nice to fully understand it.  Maybe we need to rope someone in that is an expert here.

[the remark this answers is not in the document]
Comment (MitchellGoodwin) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:540]: I see. In that case, I *believe* that because it's an internal package the unused icons will be tree shaken out. For outside font packages, if you import them it will tree shake successfully as long as you use at least [one constant from that package](https://github.com/flutter/flutter/issues/63920).

I'm unsure of what the differences would be here. It's possible that the whole icon collection would be included unnecessarily in this case, but I think only the required ones would be.

[the remark this answers is not in the document]
Comment (justinmc) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:540]: Thanks for investigating.  Hopefully the unused icons do get tree shaken out, but if not, then it's something to solve separately from this PR.  I'm sure there are many Flutter iOS apps that use TextField and want these toolbar button updates.

I think we're fine to move ahead with this PR with @tgucio's approval.

[the remark this answers is not in the document]
Comment (tgucio) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:540]: Here's what happens in a `MaterialApp` using a regular `TextField` on iOS if `cupertino_icons` isn't in the app dependencies (just swapped labels in current master with these icons and made strings longer):

toolbar

For icon tree shaked size, this example results in:
```
% ls -al build/app/intermediates/assets/release/flutter_assets/packages/cupertino_icons/assets/CupertinoIcons.ttf
-rw-r--r--@ 1 tgucio  staff  976 Jun  8 16:36 build/app/intermediates/assets/release/flutter_assets/packages/cupertino_icons/assets/CupertinoIcons.ttf
```

So size wise it's tiny but I'm still not sure about requiring the `cupertino_icons` dependency.

Maybe we can revert this part and use the former `CupertinoTextSelectionToolbarButton.text` chars - and open another PR to continue this discussion?

[the remark this answers is not in the document]
Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart]: I managed to remove the CupertinoIcons by using a `CustomPainter` to draw the chevrons.

The tests are failing because one, I didn't want to change them before asking what you think about this, and two, I want to know your opinions on what would be the best way to test the painter (probably make `_CupertinoChevronPainter` public with `@visibleForTesting`?)

[the remark this answers is not in the document]
Comment (tgucio) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart]: Good stuff! If @justinmc and @MitchellGoodwin don't have objections on this being the best approach to avoid the `cupertino_icons` dependency I'd say it's good to go.

For checking what a `CustomPainter` paints, there's a matcher `paints` with many examples e.g. in flutter/test/cupertino/context_menu_action_test.dart:
```
    await tester.pumpWidget(getApp());
    expect(find.byType(CupertinoContextMenuAction),
        paints..rect(color: kBackgroundColor.color));
```

Negative test example:
```
    expect(
      tester.firstRenderObject<RenderBox>(find.byType(CupertinoRadio<int>)),
      isNot(paints..path())
    );
```

[the remark this answers is not in the document]
Comment (justinmc) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart]: Good catch!  I didn't realize cupertino_icons would have to be explicitly included.  I'm on board with the custom painter solution :+1:

[the remark this answers is not in the document]
Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart]: @tgucio Thank you for pointing out how to test the CustomPainter, I added some tests to verify that the chevrons are pointing the correct way depending on `_CupertinoChevronPainter.isLeft` 😁

But there's another problem: in the current tests, there are lines checking if the back/next buttons are being shown, but to know this I would need a way to find a `CustomPainter` with a `_CupertinoChevronPainter` as painter and access its `isLeft` field.

The icons were found like this in the previous implementation:

```dart
  Finder findOverflowNextButton() => find.byIcon(CupertinoIcons.chevron_right);
  Finder findOverflowBackButton() => find.byIcon(CupertinoIcons.chevron_left);
```

And I could keep the same structure if `_CupertinoChevronPainter` was public, something like this would work:

```dart
  Finder findOverflowNextButton() => find.byWidgetPredicate((Widget widget) =>
    widget is CustomPaint &&
    widget.painter is CupertinoChevronPainter &&
    !((widget.painter! as CupertinoChevronPainter).isLeft,
  );
  Finder findOverflowNextButton() => find.byWidgetPredicate((Widget widget) =>
    widget is CustomPaint &&
    widget.painter is CupertinoChevronPainter &&
    ((widget.painter! as CupertinoChevronPainter).isLeft,
  );
```

But probably declaring such a specific class as public is not a good idea.

The only way I found of keeping it private but still verify back/next buttons is by checking the `runtimeType` of `CustomPainter.painter` and then, if it is `_CupertinoChevronPainter`, cast it to `dynamic` and get `isLeft`:

```dart
  Finder findOverflowNextButton() => find.byWidgetPredicate((Widget widget) =>
    widget is CustomPaint &&
    '${widget.painter?.runtimeType}' == '_CupertinoChevronPainter' &&
    !((widget.painter! as dynamic).isLeft as bool),
  );
  Finder findOverflowBackButton() => find.byWidgetPredicate((Widget widget) =>
    widget is CustomPaint &&
    '${widget.painter?.runtimeType}' == '_CupertinoChevronPainter' &&
    ((widget.painter! as dynamic).isLeft as bool),
  );
```

And even then, I don't why but I had to change all `tester.tap(findOverflowNextButton())` to `tester.tapAt(tester.getCenter(findOverflowNextButton()))` (same for the back button).

So, there's probably a better way to do this? 😅

[the remark this answers is not in the document]
Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart]: Actually, I didn't manage to test that the chevrons are pointing the correct side. The patterns are like this:

```dart
PaintPattern overflowNextPaintPattern() => paints..path(includes: <Offset>[const Offset(7.5, 5)]);
PaintPattern overflowBackPaintPattern() => paints..path(includes: <Offset>[const Offset(2.5, 5)]);
```

But even if the offsets are swapped between the two, the tests pass anyway. I think this is because `path.contains(offset)` ignores strokes and just checks if the `offset` is inside its canvas rect if there's no `fill`, and as both canvas are 10x10, both offsets always pass.

[the remark this answers is not in the document]
Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart]: Hey @justinmc @tgucio, I think you missed the notifications for my last two comments, but I would really appreciate your input here :)

File: packages/flutter/lib/src/cupertino/text_selection_toolbar.dart [hunk 13]
```diff
@@ -510,49 +526,126 @@ class _CupertinoTextSelectionToolbarContentState extends State<_CupertinoTextSel
...
+            page: _page,
+            backButton: CupertinoTextSelectionToolbarButton(
+              onPressed: _handlePreviousPage,
+              child: CustomPaint(
+                  painter: _CupertinoChevronPainter(
+                    color: _kToolbarTextColor.resolveFrom(context),
+                    iconSize: _kToolbarChevronSize,
+                    isLeft: true,
+                  ),
+                  size: const Size.fromWidth(_kToolbarChevronSize),
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:545]: This block is indented too far.

File: packages/flutter/lib/src/cupertino/text_selection_toolbar.dart (hunk 9 above)
[the remark this answers is not in the document]
Comment (justinmc) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart]: Ah sorry, how about something like this?

```dart
   ...

    Finder findOverflowButton() => find.byWidgetPredicate((Widget widget) =>
      widget is CustomPaint &&
      '${widget.painter?.runtimeType}' == '_CupertinoChevronPainter',
    );

    expect(findOverflowButton(), findsOneWidget);
    expect(
      findOverflowButton().last,
      paints
          ..path(
            includes: const <Offset>[
              Offset(7.5, 8.0),
              Offset(2.5, 13.0),
            ],
          ),
    );

    await tester.tap(findOverflowButton());
    await tester.pumpAndSettle();

    expect(findOverflowButton(), findsNWidgets(2));
    expect(
      findOverflowButton().first,
      paints
          ..path(
            includes: const <Offset>[
              Offset(2.5, 8.0),
              Offset(7.5, 13.0),
            ],
          ),
    );
    expect(
      findOverflowButton().last,
      paints
          ..path(
            includes: const <Offset>[
              Offset(7.5, 8.0),
              Offset(2.5, 13.0),
            ],
          ),
    );

```

[the remark this answers is not in the document]
Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart]: @justinmc Thank you for the suggestion :)

I don't know if `8.0` and `13.0` are the right numbers for `Offset.dy`, the chevron painter path is like this: `[(2.5, 0), (7.5, 5), (2.5, 10)]` when pointing right and `[(7.5, 0), (2.5, 5), (7.5, 10)]` when pointing left. The thing is, even tweaking this numbers like I did [here](https://github.com/flutter/flutter/pull/127757#discussion_r1231075344), it doesn't work.

Basically `paints..path()` only considers the path's `fill`, not its `stroke`, and as both the left and right chevrons have the same square area, there's no way to tell them apart with `painter..path()`.

[the remark this answers is not in the document]
Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart]: Found a way to properly test everything! I will update the tests tomorrow and commit, then you can take a look 😁

File: packages/flutter/lib/src/cupertino/text_selection_toolbar.dart (hunk 13 above)
Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart]: I changed this block to a method because it was repeated for both chevrons, now it should have the correct indentation.

File: packages/flutter/lib/src/cupertino/text_selection_toolbar.dart (hunk 9 above)
[the remark this answers is not in the document]
Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart]: I just committed, but the tl;dr is: `paints..path()` tests just the `fill`, but `paints..line()` can test the exact points that a line was drawn on, so I changed the `_CupertinoChevronPainter` to use two `canvas.drawLine`s instead of a single `canvas.drawPath`.

That made possible to test if the chevrons point to the correct side, but to actually find the chevrons I created two more private classes, one for each chevron, and now it's just a matter of comparing the `runtimeType` of the painter with either `_LeftCupertinoChevronPainter` or `_RightCupertinoChevronPainter`.

[the remark this answers is not in the document]
Comment (justinmc) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart]: Ha, I don't know why I didn't think of splitting it into two different painter classes, good call.  I'll give this another review.

**Review by justinmc — APPROVED** [review]
Renewing my LGTM with some minor comments 👍 

*Edit* I think there is a legit test failure due to trying to draw the toolbar on web, but it fails due to https://github.com/flutter/flutter/issues/44572.  If so I think you can skip it on web and reference that GitHub issue.

File: packages/flutter/lib/src/cupertino/text_selection_toolbar.dart [hunk 14]
```diff
@@ -506,53 +522,138 @@ class _CupertinoTextSelectionToolbarContentState extends State<_CupertinoTextSel
...
+// These classes help to test the chevrons. As _CupertinoChevronPainter must be
+// private, it's possible to check the runtimeType of each chevron to know if
+// they should be pointing left or right.
+class _LeftCupertinoChevronPainter extends _CupertinoChevronPainter {
+  _LeftCupertinoChevronPainter({required super.color}) : super(isLeft: true);
+}
+class _RightCupertinoChevronPainter extends _CupertinoChevronPainter {
+  _RightCupertinoChevronPainter({required super.color}) : super(isLeft: false);
+}
+class _CupertinoChevronPainter extends CustomPainter {
```
[the remark this answers is not in the document]
Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:581]: Done :)

File: packages/flutter/lib/src/cupertino/text_selection_toolbar.dart [hunk 15]
```diff
@@ -506,53 +522,138 @@ class _CupertinoTextSelectionToolbarContentState extends State<_CupertinoTextSel
...
+  final Color color;
+
+  /// If this is true the chevron will point left, else it will point right.
+  final bool isLeft;
+
+  @override
+  void paint(Canvas canvas, Size size) {
+    assert(size.height == size.width, 'size must have the same height and width');
+
+    // Height and width should always be the same.
```
[the remark this answers is not in the document]
Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar.dart]: When I added the assert I forgot to remove the comment, thanks!

File: packages/flutter/test/cupertino/text_selection_test.dart [hunk 16]
```diff
@@ -288,32 +285,29 @@ void main() {
       expect(find.text('Copy'), findsOneWidget);
       expect(find.text('Paste'), findsNothing);
       expect(find.text('Select All'), findsNothing);
-      expect(find.text('◀'), findsNothing);
-      expect(find.text('▶'), findsOneWidget);
-      expect(appearsEnabled(tester, '▶'), true);
+      expect(find.byWidgetPredicate((Widget w) => w is Icon && w.icon == CupertinoIcons.chevron_left), findsNothing);
```
[the remark this answers is not in the document]
Comment (luccasclezar) [reply packages/flutter/test/cupertino/text_selection_test.dart]: Yep 😅

File: packages/flutter/test/cupertino/text_selection_test.dart [hunk 17]
```diff
@@ -488,56 +472,48 @@ void main() {
...
       expect(find.text(_longLocalizations.selectAllButtonLabel), findsNothing);
-      expect(find.text('◀'), findsNothing);
-      expect(find.text('▶'), findsOneWidget);
-      expect(appearsEnabled(tester, '▶'), true);
+      expect(findOverflowBackButton(), findsNothing);
+      expect(findOverflowNextButton(), findsOneWidget);
 
       // Tap next to go to the second page.
-      await tester.tap(find.text('▶'));
+      await tester.tapAt(tester.getCenter(findOverflowNextButton()));
```
[the remark this answers is not in the document]
Comment (luccasclezar) [reply packages/flutter/test/cupertino/text_selection_test.dart:479]: I don't know why, but it doesn't. It complains about trying to tap on something offscreen.

**Comment by luccasclezar** [comment]
Okay then, I will skip the "chevrons point to the correct side" test on web.

But I have to say, I don't exactly understand what's the problem... It says it's on `text_selection_toolbar_test.dart:111`, but that line is the instantiation of `CupertinoTextSelectionToolbar`, which is done several other times throughout the tests, so why is this the only one failing?

File: packages/flutter/test/cupertino/text_selection_test.dart (hunk 17 above)
[the remark this answers is not in the document]
Comment (justinmc) [reply packages/flutter/test/cupertino/text_selection_test.dart:479]: Weird, I'm not sure why that would happen.  Let's leave it for now and I'll keep an eye out for this showing up again.

**Comment by justinmc** [comment]
@tgucio Can you give the final approval here?

**Review by tgucio — APPROVED** [review]

**Comment by tgucio** [comment]
LGTM - thanks again for the PR!
